File: firebase/infrastructure/functions/src/guest_replies.ts
/**
 * BLOCK-03 游客代表性回复 callable。
 *
 * 设计（PHASE-5-DISPATCH §3）：
 * - 受信服务端确定性选择：排序键 `[isVerified desc, likeCount desc, createdAt asc,
 *   replyId asc]`，无任何随机/时间戳抖动——同一数据快照 + 同一 limit/policyVersion
 *   必须返回完全相同 ID 顺序（spec Scenario 4）。
 * - 允许未认证（`request.auth == null`）：游客的完整回复读取走 callable；
 *   客户端直连 `playground_replies` 仍被 Rules 拒绝（第一道反绕过闸）。
 * - limit 服务端 clamp 到 [5,10]；忽略 cursor/page/offset/sort 等任何额外参数。
 * - 公开 DTO 不含 `author_provider_uid`/`author_app_user_id`/`presentation_identity_id`
 *   等敏感字段；tombstone 不入 visible/total。
 */

import { onCall, HttpsError } from 'firebase-functions/v2/https';
import * as crypto from 'crypto';
import { db, COLLECTIONS } from './index';

export const SELECTION_POLICY_VERSION = 1;
export const MIN_GUEST_LIMIT = 5;
export const MAX_GUEST_LIMIT = 10;

/** 展示用匿名 ID 派生：基于 provider uid 的稳定哈希，不暴露原始 uid。 */
function derivePresentationId(providerUid: unknown): string {
  const raw =
    typeof providerUid === 'string' && providerUid.length > 0
      ? providerUid
      : 'anonymous';
  return `anon_${crypto.createHash('sha256').update(raw).digest('hex').slice(0, 8)}`;
}

/** 归一化 created_at/updated_at → ISO 字符串；Timestamp/Date/string 均可。 */
function toIsoString(v: unknown): string | null {
  if (v === null || v === undefined) return null;
  if (v instanceof Date) return v.toISOString();
  if (typeof v === 'string') return v;
  if (typeof (v as any).toDate === 'function') {
    const d = (v as any).toDate();
    return d instanceof Date ? d.toISOString() : null;
  }
  return null;
}

/** 时间戳毫秒；缺失 → 最大（asc 排序时排最后），由 replyId 做最终 tie-break。 */
function toMs(v: unknown): number {
  const iso = toIsoString(v);
  if (!iso) return Number.MAX_SAFE_INTEGER;
  const ms = Date.parse(iso);
  return Number.isNaN(ms) ? Number.MAX_SAFE_INTEGER : ms;
}

/** isVerified = reply.verification 非空（verifyRootReply 写入的应验事实）。 */
function isVerified(d: Record<string, any>): boolean {
  const v = d.verification;
  return v !== null && v !== undefined && typeof v === 'object';
}

/** 公开 chart 投影：只取公开字段，忽略内部字段。 */
function toPublicChart(chart: any): any {
  if (!chart || typeof chart !== 'object') return null;
  if (chart.type !== undefined && chart.type !== 'xuanChart') return null;
  return {
    techniqueId: typeof chart.technique_id === 'string' ? chart.technique_id : 'unknown',
    schoolId: typeof chart.school_id === 'string' ? chart.school_id : null,
    publicChartSnapshot:
      typeof chart.public_chart_snapshot === 'string' ? chart.public_chart_snapshot : '',
    rendererSchemaVersion:
      typeof chart.renderer_schema_version === 'number' ? chart.renderer_schema_version : 1,
    source: typeof chart.chart_source === 'string' ? chart.chart_source : 'createdInPlayground',
  };
}

/** 公开媒体投影：只含安全 URL/token、mime、尺寸；禁止内部存储路径。 */
function toPublicMediaAttachments(list: any): any[] {
  if (!Array.isArray(list)) return [];
  return list
    .filter(
      (m) =>
        m &&
        typeof m === 'object' &&
        (m.type === 'image' || m.type === 'video'),
    )
    .map((m) => ({
      type: m.type,
      mediaObjectId:
        typeof m.media_object_id === 'string' ? m.media_object_id : 'unknown',
      mimeType: typeof m.mime_type === 'string' ? m.mime_type : '',
      width: typeof m.width === 'number' ? m.width : null,
      height: typeof m.height === 'number' ? m.height : null,
      durationSeconds:
        typeof m.duration_seconds === 'number' ? m.duration_seconds : null,
      secureUrl: `/public/media/${
        typeof m.media_object_id === 'string' ? m.media_object_id : 'unknown'
      }`,
    }));
}

/** 公开 revision 摘要：只返回最新一条元数据，不返回全文。 */
function toPublicRevisionSummary(revisions: any): any {
  if (!Array.isArray(revisions) || revisions.length === 0) return null;
  const last = revisions[revisions.length - 1];
  if (!last || typeof last !== 'object') return null;
  return {
    editedByAlias:
      typeof last.edited_by === 'string' ? last.edited_by : '未知',
    editedAt:
      toIsoString(last.edited_at) ?? '1970-01-01T00:00:00.000Z',
    changeDescription:
      typeof last.change_description === 'string' ? last.change_description : null,
  };
}

/** 公开作者投影：公开展示 ID + 展示别名；无 canonical appUserId / provider uid。 */
function toPublicAuthor(d: Record<string, any>): any {
  const presentationId =
    typeof d.presentation_identity_id === 'string' && d.presentation_identity_id.length > 0
      ? d.presentation_identity_id
      : derivePresentationId(d.author_provider_uid);
  const short = presentationId.slice(-6);
  return {
    publicPresentationUserId: presentationId,
    displayAlias: `盘友${short}`,
    avatarUrl: null,
    publicProfileRef: null,
  };
}

/** PublicReply JSON（键名 = Dart DTO 字段名约定，PHASE-5-DISPATCH §3.2）。 */
function toPublicReply(replyId: string, d: Record<string, any>): any {
  return {
    publicReplyId: replyId,
    postId: typeof d.post_id === 'string' ? d.post_id : '',
    body: typeof d.body === 'string' ? d.body : '',
    techniqueTags: Array.isArray(d.technique_tags) ? d.technique_tags : [],
    chart: toPublicChart(d.chart_attachment),
    mediaAttachments: toPublicMediaAttachments(d.media_attachments),
    author: toPublicAuthor(d),
    depth: typeof d.depth === 'number' ? d.depth : 0,
    rootReplyId: typeof d.root_reply_id === 'string' ? d.root_reply_id : null,
    replyToReplyId:
      typeof d.reply_to_reply_id === 'string' ? d.reply_to_reply_id : null,
    isVerified: isVerified(d),
    isTombstoned: false,
    presentationMode: 'stableAlias',
    createdAt: toIsoString(d.created_at) ?? '1970-01-01T00:00:00.000Z',
    updatedAt: toIsoString(d.updated_at),
    latestRevision: toPublicRevisionSummary(d.revisions),
  };
}

/** 取该 post 最新的有效最终反馈（deleted_at==null）；无 → null。游客保持可见。 */
async function loadOutcomeFeedback(postId: string): Promise<any | null> {
  const snaps = await db
    .collection(COLLECTIONS.outcomeFeedback)
    .where('post_id', '==', postId)
    .get();
  const active = snaps.docs
    .map((doc) => ({ id: doc.id, data: doc.data() ?? {} }))
    .filter((x) => x.data.deleted_at === null || x.data.deleted_at === undefined);
  if (active.length === 0) return null;

  active.sort((a, b) => {
    const aMs = toMs(a.data.created_at);
    const bMs = toMs(b.data.created_at);
    if (aMs !== bMs) return bMs - aMs; // 最新在前
    return a.id < b.id ? 1 : a.id > b.id ? -1 : 0;
  });

  const latest = active[0].data;
  const publishedAt =
    toIsoString(latest.created_at) ?? '1970-01-01T00:00:00.000Z';
  const updatedAt = toIsoString(latest.updated_at);
  return {
    body:
      typeof latest.outcome_description === 'string'
        ? latest.outcome_description
        : '',
    isEdited: updatedAt !== null && updatedAt !== publishedAt,
    publishedAt,
    updatedAt,
  };
}

export const getGuestRepresentativeReplies = onCall(
  { region: 'asia-east1' },
  async (request) => {
    // 游客 callable：允许 request.auth == null（不调用 requireAuthUid）。

    const data = (request.data ?? {}) as Record<string, any>;

    // 入参白名单：只取 postId / limit / selectionPolicyVersion；
    // cursor/page/offset/sort 等任何额外参数一律忽略（反绕过）。
    const postId = data.postId;
    if (typeof postId !== 'string' || postId.trim().length === 0) {
      throw new HttpsError('invalid-argument', 'postId 不能为空');
    }

    // limit：服务端 clamp 到 [5,10]；非法/缺省 → 5。
    let limit = MIN_GUEST_LIMIT;
    if (typeof data.limit === 'number' && Number.isInteger(data.limit)) {
      limit = Math.min(MAX_GUEST_LIMIT, Math.max(MIN_GUEST_LIMIT, data.limit));
    }

    // selectionPolicyVersion：仅支持 1；未知版本 → invalid-argument；缺省 → 1。
    const policyVersion =
      data.selectionPolicyVersion === undefined ||
      data.selectionPolicyVersion === null
        ? SELECTION_POLICY_VERSION
        : data.selectionPolicyVersion;
    if (
      typeof policyVersion !== 'number' ||
      policyVersion !== SELECTION_POLICY_VERSION
    ) {
      throw new HttpsError(
        'invalid-argument',
        `不支持的 selectionPolicyVersion: ${String(policyVersion)}`,
      );
    }

    const postSnap = await db.collection(COLLECTIONS.posts).doc(postId).get();
    if (!postSnap.exists || postSnap.get('status') !== 'active') {
      throw new HttpsError('not-found', '帖子不存在或已失效');
    }

    // root replies（depth==0）；tombstone 在内存过滤，不入 total/visible。
    const rootsSnap = await db
      .collection(COLLECTIONS.replies)
      .where('post_id', '==', postId)
      .where('depth', '==', 0)
      .get();

    const candidates: Array<{ id: string; data: Record<string, any>; createdAtMs: number }> =
      [];
    for (const doc of rootsSnap.docs) {
      const d = doc.data() ?? {};
      if (d.is_tombstoned === true) continue;
      candidates.push({ id: doc.id, data: d, createdAtMs: toMs(d.created_at) });
    }

    // likes 计数：playground_likes 按 reply_id count（真实 Admin SDK 查询）。
    const likeCounts = new Map<string, number>();
    for (const c of candidates) {
      const likeSnap = await db
        .collection(COLLECTIONS.likes)
        .where('reply_id', '==', c.id)
        .get();
      likeCounts.set(c.id, likeSnap.size);
    }

    // 确定性选择（policy v1）：
    // [isVerified desc, likeCount desc, createdAt asc, replyId asc]。
    const sorted = [...candidates].sort((a, b) => {
      const aVerified = isVerified(a.data);
      const bVerified = isVerified(b.data);
      if (aVerified !== bVerified) return aVerified ? -1 : 1;
      const aLikes = likeCounts.get(a.id) ?? 0;
      const bLikes = likeCounts.get(b.id) ?? 0;
      if (aLikes !== bLikes) return bLikes - aLikes;
      if (a.createdAtMs !== b.createdAtMs) return a.createdAtMs - b.createdAtMs;
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });

    const visible = sorted.slice(0, Math.min(limit, sorted.length));
    const totalReplyCount = candidates.length;
    const hiddenReplyCount = Math.max(0, totalReplyCount - visible.length);

    const outcomeFeedback = await loadOutcomeFeedback(postId);

    return {
      postId,
      visibleReplies: visible.map((c) => toPublicReply(c.id, c.data)),
      totalReplyCount,
      hiddenReplyCount,
      selectionPolicyVersion: policyVersion,
      registrationUnlock: {
        requiresRegistration: true,
        ctaMessageKey: 'register_to_unlock_replies',
        unlockRoute: '/register',
      },
      outcomeFeedback,
    };
  },
);

File: firebase/infrastructure/functions-py/xuan/handlers/guest_replies.py
"""游客代表性回复。基准实现：functions/src/guest_replies.ts

★ 本模块是唯一允许未认证调用的 callable —— 游客读完整回复走这里，
客户端直连 playground_replies 仍被 Rules 拒绝。
因此 `_impl` 的签名里**没有 uid**，不要加。
"""

import json

from firebase_functions import https_fn

from xuan.cache import (
    cached_query,
    compute_query_fingerprint,
    get_global_cache,
    get_global_single_flight,
)
from xuan.config import COLLECTIONS, REGION, db
from xuan.errors import invalid_argument, not_found
from xuan.public_dto import (
    is_real_number, is_verified, to_iso_string, to_ms, to_public_reply,
)

SELECTION_POLICY_VERSION = 1
MIN_GUEST_LIMIT = 5
MAX_GUEST_LIMIT = 10


def _load_outcome_feedback(post_id: str):
    """取该帖最新的有效最终反馈（deleted_at 为空）。无 → None。"""
    client = db()
    snaps = client.collection(COLLECTIONS["outcome_feedback"]).where("post_id", "==", post_id).get()
    active = [(d.id, d.to_dict() or {}) for d in snaps]
    active = [x for x in active if x[1].get("deleted_at") is None]
    if not active:
        return None

    # 最新在前；同一时刻用 id **降序** tie-break（★ 坑 5：与主排序方向相反）
    active.sort(key=lambda x: (to_ms(x[1].get("created_at")), x[0]), reverse=True)
    latest = active[0][1]

    published_at = to_iso_string(latest.get("created_at")) or "1970-01-01T00:00:00.000Z"
    updated_at = to_iso_string(latest.get("updated_at"))
    return {
        "body": latest["outcome_description"] if isinstance(latest.get("outcome_description"), str) else "",
        "isEdited": updated_at is not None and updated_at != published_at,
        "publishedAt": published_at,
        "updatedAt": updated_at,
    }


def _get_guest_representative_replies_impl(data: dict) -> dict:
    """★ 无 uid 参数：游客可调。

    已接入 CachePort 读缓存网关（查询指纹 + 单飞防击穿 + TTL 抖动 + ETag 304 短路）。
    """
    data = data or {}

    post_id = data.get("postId")
    if not isinstance(post_id, str) or not post_id.strip():
        raise invalid_argument("postId 不能为空")

    # limit：只接受真整数（★ 坑 1：必须排除 bool），clamp 到 [5,10]，否则回落 5
    limit = MIN_GUEST_LIMIT
    raw_limit = data.get("limit")
    if is_real_number(raw_limit) and float(raw_limit).is_integer():
        limit = min(MAX_GUEST_LIMIT, max(MIN_GUEST_LIMIT, int(raw_limit)))

    # policyVersion：仅支持 1；缺省/None → 1；其余一律非法（bool 同样非法）
    raw_version = data.get("selectionPolicyVersion")
    policy_version = SELECTION_POLICY_VERSION if raw_version is None else raw_version
    if not is_real_number(policy_version) or policy_version != SELECTION_POLICY_VERSION:
        raise invalid_argument(f"不支持的 selectionPolicyVersion: {policy_version}")

    cache_key = compute_query_fingerprint(
        route=f"guest_replies/{post_id}",
        params={"limit": limit, "selectionPolicyVersion": policy_version},
        contract_version="1",
        default_params={"limit": MIN_GUEST_LIMIT, "selectionPolicyVersion": SELECTION_POLICY_VERSION},
    )

    if_none_match = data.get("ifNoneMatch") or data.get("if_none_match")

    def _loader() -> dict:
        client = db()
        post_snap = client.collection(COLLECTIONS["posts"]).document(post_id).get()
        if not post_snap.exists or (post_snap.to_dict() or {}).get("status") != "active":
            raise not_found("帖子不存在或已失效")

        roots = client.collection(COLLECTIONS["replies"]) \
            .where("post_id", "==", post_id).where("depth", "==", 0).get()

        candidates = []
        for doc in roots:
            d = doc.to_dict() or {}
            if d.get("is_tombstoned") is True:
                continue   # tombstone 不入 total，也不入 visible
            candidates.append((doc.id, d, to_ms(d.get("created_at"))))

        # 批量查询 likes（Firestore 'in' 限制单批最多 30 个），并在内存中归组计数
        # 保证 0 赞的 reply_id 仍然存在且计数为 0
        like_counts = {cid: 0 for cid, _, _ in candidates}
        cids = [cid for cid, _, _ in candidates]
        batch_size = 30
        for i in range(0, len(cids), batch_size):
            chunk = cids[i:i + batch_size]
            if not chunk:
                continue
            likes_docs = client.collection(COLLECTIONS["likes"]).where("reply_id", "in", chunk).get()
            for doc in likes_docs:
                d = doc.to_dict() or {}
                rid = d.get("reply_id")
                if rid in like_counts:
                    like_counts[rid] += 1

        # ★ 坑 3：[isVerified desc, likeCount desc, createdAt asc, replyId asc]
        ordered = sorted(candidates, key=lambda c: (
            0 if is_verified(c[1]) else 1,   # 已验在前
            -like_counts[c[0]],              # 赞多在前
            c[2],                            # 早的在前
            c[0],                            # 最终 tie-break，保证确定性
        ))

        visible = ordered[:min(limit, len(ordered))]
        total = len(candidates)

        return {
            "postId": post_id,
            "visibleReplies": [to_public_reply(cid, d) for cid, d, _ in visible],
            "totalReplyCount": total,
            "hiddenReplyCount": max(0, total - len(visible)),
            "selectionPolicyVersion": policy_version,
            "registrationUnlock": {
                "requiresRegistration": True,
                "ctaMessageKey": "register_to_unlock_replies",
                "unlockRoute": "/register",
            },
            "outcomeFeedback": _load_outcome_feedback(post_id),
        }

    cached_resp = cached_query(
        cache=get_global_cache(),
        key=cache_key,
        loader=_loader,
        if_none_match=if_none_match,
        ttl=60.0,
        single_flight=get_global_single_flight(),
    )

    if cached_resp.status_code == 304:
        return {"_status_code": 304, "_etag": cached_resp.etag}

    if isinstance(cached_resp.body, (bytes, bytearray)):
        res_dict = json.loads(cached_resp.body.decode("utf-8"))
    elif isinstance(cached_resp.body, str):
        res_dict = json.loads(cached_resp.body)
    else:
        res_dict = cached_resp.body

    if cached_resp.etag:
        res_dict["_etag"] = cached_resp.etag
    return res_dict


@https_fn.on_call(region=REGION)
def get_guest_representative_replies_py(req: https_fn.CallableRequest) -> dict:
    """游客读代表性回复。**不校验登录**，req.auth 可为 None。"""
    return _get_guest_representative_replies_impl(req.data or {})
